File: src/test/java/backend/controller/TestPostController.java
package backend.controller;

import org.junit.Test;
import org.springframework.http.MediaType;

import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

public class TestPostController extends backend.controller.IntegrationTest {

    @Test
    public void createNewPost() throws Exception {
        //TODO: Check whether Post was really created and saved to database


        //  {
        //      "created": 1388530800000,  // UNIX Timestamp: Wed Jan 01 2014 00:00:00 GMT+0100 (CET)
        //                                 // Date of creation in App, time when user hits "POST" button
        //
        //      "sent": 1388617200000,     // Unix Timestamp: Thu Jan 02 2014 00:00:00 GMT+0100 (CET)
        //                                 // Time when post leaves queue of app, in case app was offline, may be in HTTP post??
        //      "text": "String",          // Content of post
        //      "location" : {
        //          "lat": 51.3,
        //          "lon": 17.2,
        //      },
        //      "challenge_id" : "rand_id" // ID of a challenge that thereby get's fulfilled by the users team!
        //  }
        String json = "{\"created\": 1388530800000,\"sent\": 1388617200000,\"text\": \"String\",\"location\" : {\"lat\": 51.3,\"lon\": 17.2},\"challenge_id\" : \"rand_id\"}";
        mockMvc.perform(post("/test/post/", json))
                .andExpect(status().isCreated())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(jsonPath("id").exists());

    }

    @Test
    public void dontCreatePostForInvalidJSON() throws Exception {
        String missingField = "{\"sent\": 1388617200000,\"text\": \"String\",\"location\" : {\"lat\": 51.3,\"lon\": 17.2},\"challenge_id\" : \"rand_id\"}";
        String missingNestedField = "{\"created\":123456,\"sent\": 1388617200000,\"text\": \"String\",\"location\" : {\"lon\": 17.2},\"challenge_id\" : \"rand_id\"}";
        String invalidJSON = "{#;}\"created\": 1388530800000,\"sent\": 1388617200000,\"text\": \"String\",\"location\" : {\"lat\": 51.3,\"lon\": 17.2},\"challenge_id\" : \"rand_id\"}";

        mockMvc.perform(post("/test/post/", missingField))
                .andExpect(status().isBadRequest());

        mockMvc.perform(post("/test/post/", invalidJSON))
                .andExpect(status().isBadRequest());

        mockMvc.perform(post("/test/post/", missingNestedField))
                .andExpect(status().isBadRequest());
    }
}

File: src/test/java/backend/controller/TestPostController.kt
@file:JvmName("TestPostController")

package backend.controller

import backend.Integration.Credentials
import backend.Integration.IntegrationTest
import backend.Integration.createUser
import backend.Integration.toJsonString
import org.junit.Before
import org.junit.Ignore
import org.junit.Test
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders

import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import java.time.LocalDateTime
import java.time.ZoneOffset
import kotlin.collections.forEach
import kotlin.collections.mapOf

class TestPostController : IntegrationTest() {

    lateinit var userCredentials: Credentials

    @Before
    override fun setUp() {
        super.setUp()
        userCredentials = createUser(this.mockMvc)
    }

    @Test
    @Throws(Exception::class)
    fun createNewPost() {
        val postData = mapOf(
                "text" to "TestPost",
                "date" to LocalDateTime.now().toEpochSecond(ZoneOffset.UTC),
                "postLocation" to mapOf(
                        "latitude" to 0.0,
                        "longitude" to 0.0
                )
        ).toJsonString()

        val request = MockMvcRequestBuilders
                .request(HttpMethod.POST, "/post/")
                .header("Authorization", "Bearer ${userCredentials.accessToken}")
                .contentType(MediaType.APPLICATION_JSON)
                .content(postData)

        val response = mockMvc.perform(request)
                .andExpect(status().isCreated)
                .andExpect(content().contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(jsonPath("$.id").exists())
                .andExpect(jsonPath("$.text").exists())
                .andExpect(jsonPath("$.date").exists())
                .andExpect(jsonPath("$.user").exists())
                .andExpect(jsonPath("$.postLocation.latitude").exists())
                .andExpect(jsonPath("$.postLocation.longitude").exists())
                .andReturn().response.contentAsString

        println(response)
    }

    @Test
    @Throws(Exception::class)
    fun dontCreatePostForInvalidJSON() {
        val postData = mapOf(
                "date" to LocalDateTime.now().toEpochSecond(ZoneOffset.UTC),
                "postLocation" to mapOf(
                        "latitude" to 0.0,
                        "longitude" to 0.0
                )
        ).toJsonString()

        val request = MockMvcRequestBuilders
                .request(HttpMethod.POST, "/post/")
                .header("Authorization", "Bearer ${userCredentials.accessToken}")
                .contentType(MediaType.APPLICATION_JSON)
                .content(postData)

        val response = mockMvc.perform(request)
                .andExpect(status().isBadRequest)
                .andReturn().response.contentAsString

        println(response)
    }


    @Test
    @Throws(Exception::class)
    fun dontCreatePostWithoutValidAuth() {
        val postData = mapOf(
                "date" to LocalDateTime.now().toEpochSecond(ZoneOffset.UTC),
                "postLocation" to mapOf(
                        "latitude" to 0.0,
                        "longitude" to 0.0
                )
        ).toJsonString()

        val request = MockMvcRequestBuilders
                .request(HttpMethod.POST, "/post/")
                .header("Authorization", "Bearer invalidToken")
                .contentType(MediaType.APPLICATION_JSON)
                .content(postData)

        val response = mockMvc.perform(request)
                .andExpect(status().isUnauthorized)
                .andReturn().response.contentAsString

        println(response)
    }
}